src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,346 @@
+            {
+                FrameworkErrorUtilities.VerifyThrow(_capacity != -1, "Reusing after dispose");
+
+                _borrowedBuilder = ReuseableStringBuilderFactory.Get(_capacity);
+                _borrowedWithCapacity = _borrowedBuilder.Capacity;
+            }
+        }
+
+        /// <summary>
+        /// A utility class that mediates access to a shared string builder.
+        /// </summary>
+        /// <remarks>
+        /// If this shared builder is highly contended, this class could add
+        /// a second one and try both in turn.
+        /// </remarks>
+        private static class ReuseableStringBuilderFactory
+        {
+            /// <summary>
+            /// Made up limit beyond which we won't share the builder
+            /// because we could otherwise hold a huge builder indefinitely.
+            /// This was picked empirically to save at least 95% of allocated data size.
+            /// This constant has to exactly 2^n (power of 2) where n = 4 ... 32
+            /// </summary>
+            /// <remarks>
+            /// This constant might looks huge, but rather that lowering this constant,
+            ///   we shall focus on eliminating of code which requires to create such huge strings.
+            /// </remarks>
+            private const int MaxBuilderSizeBytes = 2 * 1024 * 1024; // ~1M chars
+            private const int MaxBuilderSizeCapacity = MaxBuilderSizeBytes / 2;
+
+            private static readonly IReadOnlyList<int> s_capacityBrackets;
+
+            static ReuseableStringBuilderFactory()
+            {
+                var brackets = new List<int>();
+
+                int bytes = 0x200; // Minimal capacity is 256 (512 bytes) as this was, according to captured traces, mean required capacity
+                while (bytes <= MaxBuilderSizeBytes)
+                {
+                    // Allocation of arrays is optimized in byte[bytes] => bytes = 2^n.
+                    // StringBuilder allocates chars[capacity] and each char is 2 bytes so lets have capacity brackets computed as `bytes/2` 
+                    brackets.Add(bytes/2); 
+                    bytes <<= 1;
+                }
+                Debug.Assert((bytes >> 1) == MaxBuilderSizeBytes, "MaxBuilderSizeBytes has to be 2^n (power of 2)");
+
+                s_capacityBrackets = brackets;
+            }
+
+            /// <summary>
+            /// The shared builder.
+            /// </summary>
+            private static StringBuilder s_sharedBuilder;
+
+#if DEBUG && ASSERT_BALANCE
+            /// <summary>
+            /// Balance between calling Get and Release.
+            /// Shall be always 0 as Get and 1 at Release.
+            /// </summary>
+            private static int s_getVsReleaseBalance;
+#endif
+
+            /// <summary>
+            /// Obtains a string builder which may or may not already
+            /// have been used. 
+            /// Never returns null.
```

Comment:
`#nullable enable` would be better than capturing nullness in comments.

Reply:
done